### ISSUE TITLE
Better support for pypip packages implementing torch cuda extentions

### ISSUE THREAD

**Issue by charliebudd** [pytorch/pytorch#72712]
### 🚀 The feature, motivation and pitch

Recently I have been trying to publish a pypip package which implements a cuda based algorithm for use in pytorch. The ideal result would be that the user simply runs `pip install <package-name>`. However, as of torch 1.10, the code needs to be compiled by the same cuda version that was used to compile the users torch install. This makes it harder for developers to publish packages as they need to provide wheels for each cuda version they want to support.  It also complicates the install process `pip install <package-name>-<torch-cuda-version>`. 

One possible solution would be to include the correct nvcc compiler (and other required components) with the cuda binaries already included with the torch distibutions. A change to CUDAExtention could then be made to use the version of nvcc included in the pytorch install.

### Alternatives

Exisitng extentions, [pytorch-geometric](https://pytorch-geometric.readthedocs.io/en/latest/notes/installation.html) for example, publish binary distributions for each cuda version. This is not ideal as it requires extra work to deploy packages and makes it more complex to install.

I found [nvidias pyindex](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#pip-wheels-installation-linux) where they publish various cuda runtimes for installation via pip. This includes `nvidia-cuda-nvcc-<cuda-version>` however it does not include the nvcc executable so solutions where this is installed as a build requirement wont work. (As a side note, would these runtimes be a better solution to pytorches current manual inclusion of cuda runtimes in the pip package?)

**Comment by malfet** [comment]
Yes, we can not redistribute NVCC, but you can compile your kernels into a PTX format, which later can be jit-compiled on host machine as long as CUDA runtime is installed.It might, in some cases be not as performant as compilation directly into CUDA machine code, but would solve the problem. Also, to the best of my knowledge, CUDA is built with the idea that different versions of runtime can co-exist in the same address space, i.e. one should be able to install pytorch compiled with CUDA-11.3 and say extension compiled with CUDA-11.5

**Comment by charliebudd** [comment]
Yes I think the most recent version of cpp_extention.py only checks the major version with just a warning if the minor version is wrong. Still would be nice to not have to have cuda installed like with the main torch installation. 

Interesting, so going from PTX to a compiled program doesn't require nvcc? In which case could it be built into CUDAExtention to compile the ptx files when building a package for distribution and use this during install?

**Comment by charliebudd** [comment]
It seems that the most elegant solution at the moment is to publish tagged versions `<package-name>==<version-number>+cu<cuda-version>` and have the user install the version which matches the cuda runtime installed with their pytorch install. 

Setting this up robustly is a non-trivial task and, seeing as torch has already tackled these/similar problems, I think a good solution would be to publish a series of docker images created for building torch extentions for all main distros of pytorch. It would be nice to wrap the process in a github action hosted in the main pytorch repository so developers can easily include this in there release process.

**Comment by charliebudd** [comment]
Related issue on the pytorch builder repo... https://github.com/pytorch/builder/issues/468

**Comment by charliebudd** [comment]
So I went ahead and solved this in what I think is a suitable way. I now host an open source github workflow [here](https://github.com/charliebudd/torch-extension-builder). Devs can call the workflow from there own CICD pipelines. The workflow then builds the code for each version of CUDA and PyTorch currently supported. These then get slimmed down by removing dependancies from the wheel that will be present in the torch/lib folder of the end users torch install. Finally these wheels are combined so that each wheel contains all compilations. The compilation to use is dynamically loaded at runtime depending on the version of torch the end users has installed.
